File: localhub-frontend/src/api/posts.js
import { apiFetch } from './client'

export function listPosts({
  category,
  keyword,
  sort = 'latest',
  page = 1,
  size = 10,
  locationId,
} = {}) {
  const params = new URLSearchParams()
  if (category) params.set('category', category)
  if (keyword) params.set('keyword', keyword)
  if (locationId) params.set('location_id', locationId)
  params.set('sort', sort)
  params.set('page', String(page))
  params.set('size', String(size))
  return apiFetch(`/api/posts?${params.toString()}`)
}

export function getPost(postId) {
  return apiFetch(`/api/posts/${postId}`)
}

export function createPost(payload) {
  return apiFetch('/api/posts', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function updatePost(postId, payload) {
  return apiFetch(`/api/posts/${postId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

export function deletePost(postId, password) {
  return apiFetch(`/api/posts/${postId}/delete`, {
    method: 'POST',
    body: JSON.stringify({ password }),
  })
}

export function createComment(postId, { content, password }) {
  return apiFetch(`/api/posts/${postId}/comments`, {
    method: 'POST',
    body: JSON.stringify({ content, password }),
  })
}

export function deleteComment(commentId, password) {
  return apiFetch(`/api/posts/comments/${commentId}/delete`, {
    method: 'POST',
    body: JSON.stringify({ password }),
  })
}

const LIST_ALL_PAGE_SIZE = 50
const LIST_ALL_MAX_PAGES = 20 // 최대 1,000건까지 — 게시판 전체를 클라이언트에서 필터/정렬하기 위한 상한선

// 게시판은 카테고리 다중 선택 + 키워드 검색을 클라이언트에서 처리하므로, 전체 목록을 이어서 가져온다.
export async function listAllPosts() {
  const first = await listPosts({ page: 1, size: LIST_ALL_PAGE_SIZE })
  const totalPages = Math.min(
    Math.ceil(first.total / LIST_ALL_PAGE_SIZE) || 1,
    LIST_ALL_MAX_PAGES,
  )

  const restPages = []
  for (let page = 2; page <= totalPages; page += 1) restPages.push(page)

  const rest = await Promise.all(
    restPages.map((page) => listPosts({ page, size: LIST_ALL_PAGE_SIZE })),
  )

  return [first.items, ...rest.map((response) => response.items)].flat()
}

File: localhub-backend/app/routers/posts.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app import crud, schemas
from app.database import get_db

router = APIRouter(prefix="/api/posts", tags=["posts"])


@router.get("", response_model=schemas.PostListResponse)
def list_posts(
    category: str | None = None,
    keyword: str | None = None,
    sort: str = Query("latest", pattern="^(latest|rating|views)$"),
    page: int = Query(1, ge=1),
    size: int = Query(10, ge=1, le=50),
    location_id: str | None = None,
    db: Session = Depends(get_db),
):
    """게시글 목록 (카테고리 필터 / 검색 / 정렬 / 페이지네이션 / 장소별 리뷰 필터)."""
    total, items = crud.get_posts(db, category, keyword, sort, page, size, location_id)
    return {"total": total, "page": page, "size": size, "items": items}


@router.get("/{post_id}", response_model=schemas.PostDetail)
def get_post(post_id: int, db: Session = Depends(get_db)):
    """상세 조회 (조회수 +1)."""
    post = crud.get_post_and_increment_view(db, post_id)
    if not post:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "게시글을 찾을 수 없습니다.")
    return post


@router.post("", response_model=schemas.PostDetail, status_code=201)
def create_post(data: schemas.PostCreate, db: Session = Depends(get_db)):
    """게시글 작성 (제목·내용·비밀번호·평점)."""
    return crud.create_post(db, data)


@router.put("/{post_id}", response_model=schemas.PostDetail)
def update_post(
    post_id: int, data: schemas.PostUpdate, db: Session = Depends(get_db)
):
    """수정 — 비밀번호 평문 일치 확인 (RFP 의도된 설계)."""
    post = crud.get_post(db, post_id)
    if not post:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "게시글을 찾을 수 없습니다.")
    if post.password != data.password:
        raise HTTPException(status.HTTP_403_FORBIDDEN, "비밀번호가 일치하지 않습니다.")
    return crud.update_post(db, post, data)


@router.post("/{post_id}/delete", status_code=204)
def delete_post(
    post_id: int, body: schemas.PasswordCheck, db: Session = Depends(get_db)
):
    """
    삭제 — 비밀번호 확인.
    DELETE 대신 POST /delete 를 쓰는 이유: DELETE 메서드는 body 전송이
    클라이언트/프록시에 따라 무시될 수 있어, 비밀번호를 안전히 보내기 위함.
    """
    post = crud.get_post(db, post_id)
    if not post:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "게시글을 찾을 수 없습니다.")
    if post.password != body.password:
        raise HTTPException(status.HTTP_403_FORBIDDEN, "비밀번호가 일치하지 않습니다.")
    crud.delete_post(db, post)
